### PR TITLE
libcore => 2018

### PR DESCRIPTION
**Pull request by taiki-e**
Transitions `libcore` to Rust 2018; cc #58099

r? @Centril

### CODE REVIEW

**Comment by gnzlbg** [comment]
This is probably blocked on https://github.com/rust-lang/rust/pull/58675

**Comment by taiki-e** [comment]
```text
[00:04:33] error[E0433]: failed to resolve: could not find `ops` in `{{root}}`
[00:04:33] 
[00:04:33] error[E0433]: failed to resolve: could not find `option` in `{{root}}`
[00:04:33] 
[00:04:33] error[E0433]: failed to resolve: could not find `iter` in `{{root}}`
[00:04:33] 
[00:04:33] error[E0433]: failed to resolve: could not find `result` in `{{root}}`
[00:04:33] 
[00:04:33] error[E0433]: failed to resolve: could not find `convert` in `{{root}}`
```

I don't know where these errors came from...

**Comment by Centril** [comment]
@taiki-e Could you try to limit this PR to just the bare minimum that is needed to get core into 2018? Maybe that will weed out the cause of the failure...

**Comment by taiki-e** [comment]
Blocked on #59462

**Comment by Centril** [comment]
https://github.com/rust-lang/rust/pull/59462 has landed. Rebase?

**Comment by taiki-e** [comment]
@Centril Probably need to change the bootstrap.

**Comment by Centril** [comment]
@taiki-e The bootstrap compiler will get updated after we promote master to beta.

**Comment by Centril** [comment]
@taiki-e Want to give rebasing another try now that bootstrap was updated?

**Comment by mati865** [comment]
@Centril I think it's still old beta: https://github.com/rust-lang/rust/blob/60076bb8f7e7ca7fc98ca432cda86dbdce7c0417/src/stage0.txt#L15

**Comment by Centril** [comment]
@pietroalbini huh, I thought we had bumped boostrap...?

**Comment by mati865** [comment]
Looking at PRs only stable and beta bootstraps were updated so far.

**Comment by Centril** [comment]
@mati865 Yep; seems right -- would you like to file a bootstrap update PR for master?

**Comment by mati865** [comment]
I was wondering about doing it already (for different issue) but I thought I'd be better to leave it to the team.

File: src/libcore/lib.rs [hunk 1]
```diff
@@ -62,6 +62,9 @@
 #![warn(missing_docs)]
 #![warn(intra_doc_link_resolution_failure)]
 #![warn(missing_debug_implementations)]
+#![deny(rust_2018_idioms)]
```
Comment (RalfJung) [inline src/libcore/lib.rs]: For consistency with the other crates, could you add an empty line above this?

(Yes I know I have an issue if I care about this.^^)

**Comment by taiki-e** [comment]
(Rebase for resolve conflicts and https://github.com/rust-lang/rust/pull/58702#discussion_r275167387)

**Comment by gnzlbg** [comment]
I think this is blocking `stdsimd` from using `edition = 2018`, do not stress yourselves because of this, but it would be awesome if you could open an issue in the `stdsimd` repo once this is merged to make the switch there. Cheers!

**Comment by Centril** [comment]
@taiki-e Bootstrap compiler was just updated in https://github.com/rust-lang/rust/pull/59974.

**Comment by taiki-e** [comment]
Rebased!

**Comment by Centril** [comment]
cc @gnzlbg ^---

**Comment by gnzlbg** [comment]
So I've opened https://github.com/rust-lang-nursery/stdsimd/pull/725 to set the edition of `core::arch` to 2018 unconditionally, and will hold all contributions to `stdsimd` until that PR and this PR are merged, because ensuring that the `core::arch` module remains compatible for both editions for all tier1 and tier2 targets is just too much of a pain.

**Comment by Centril** [comment]
Yay!

**Comment by Centril** [comment]
I'll prepare a PR for the `#[allow(...)]` bits :)

**Pull request merged**